server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,416 @@
+                                    tracing::error!("{} keyring incorrect secret.", &label);
+
+                                    let properties = Properties::for_unlock(
+                                        &label,
+                                        "The unlock password was incorrect",
+                                        &self.window_id,
+                                    );
+                                    let path = self.path.clone();
+
+                                    let server_exchange =
+                                        self.service.secret_exchange_pub_key().await;
+
+                                    tokio::spawn(PrompterCallback::perform_prompt(
+                                        connection.clone(),
+                                        path,
+                                        PromptType::Password,
+                                        properties,
+                                        server_exchange,
+                                    ));
+
+                                    return Ok(());
+                                }
+                                Err(err) => {
+                                    return Err(ServiceError::ZBus(zbus::Error::FDO(Box::new(
+                                        zbus::fdo::Error::Failed(format!(
+                                            "Failed to unlock {label} Keyring: {err}."
+                                        )),
+                                    ))))
+                                }
+                            }
+                        }
+                    }
+
+                    let service = self.service.clone();
+                    let result = Value::new(objects.clone()).try_to_owned().unwrap();
+
+                    tokio::spawn(async move {
+                        let _ = service.set_locked(locked, &objects, true).await;
+                    });
+
+                    tokio::spawn(PrompterCallback::stop_prompting(
+                        connection,
+                        self.path.clone(),
+                    ));
+
+                    let signal_emitter = self.service.signal_emitter(prompt_path)?;
+
+                    tokio::spawn(PrompterCallback::prompt_completed(
+                        signal_emitter,
+                        false,
+                        result,
+                    ));
+                }
+            }
+        } else {
+            // First PromptReady call
+            let secret_exchange = SecretExchange::new().map_err(|err| {
+                ServiceError::ZBus(zbus::Error::FDO(Box::new(zbus::fdo::Error::Failed(
+                    format!("Failed to generate SecretExchange {err}."),
+                ))))
+            })?;
+            let server_exchange = secret_exchange.begin();
+            self.service
+                .set_secret_exchange_pub_key(server_exchange.clone())
+                .await;
+
```

Comment:
why this information is stored in the service and not in the prompter itself? i don't think the gnome specific bits needs to be exposed in Service.

Reply:
Yeah, I agree. Removed these from the Service.